src/core/drive/preloader.js

```diff
@@ -5,6 +5,13 @@ export class Preloader {
 
   constructor(delegate) {
     this.delegate = delegate
+    this.preloadLinkObserver = new MutationObserver((mutationList, observer) => {
+      mutationList.forEach((mutation) => {
+        if (mutation.attributeName !== "data-turbo-preload" || mutation.target.tagName !== "A") return
+
+        this.preloadURL(mutation.target)
+      })
+    })
```

Comment:
While I appreciate that this PR's code matches the style of the surrounding code, I wonder if the introduction of the `MutationObserver` poses an opportunity to re-structure how the Preloader monitors the page.

There's some precedent for code tied to the document's structure, nodes, and attributes to live in the [src/observers](https://github.com/hotwired/turbo/tree/main/src/observers) directory as a class with a `.start()` and `.stop()` method and a context-specific delegate interface (when the project was built with TypeScript, enforcing that delegate interface was fairly automatic).

In this case, calling `.start()` could attach the `MutationObserver` (maybe even with [attributeFilter: "data-turbo-preload"](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver/observe)), and calling `.stop()` could detach it.

I'm imagining a `PreloadObserver` resembling something like:

```js
class PreloadObserver {
  started = false
  attributeName = "data-turbo-preload"

  constructor(delegate, element) {
    this.delegate = delegate
    this.element = element
    this.mutationObserver = new MutationObserver(this.#notifyDelegateOfMutations)
  }

  start() {
    if (!this.started) {
      this.mutationObserver.observe(this.element, {
        attributeFilter: [this.attributeName],
        subtree: true
      })

      for (const element of this.element.querySelectorAll(`[${this.attributeName}]`)) {
        this.#notifyDelegate(element)
      }
  
      this.started = true
    }
  }

  stop() {
    if (this.started) {
      this.mutationObserver.disconnect()
      this.started = false
    }
  }

  #notifyDelegateOfMutations = (mutationList) => {
    mutationList.forEach(({ target }) => this.#notifyDelegate(target))
  }

  #notifyDelegate = (element) => {
    if (element instanceof HTMLAnchorElement) {
      // this is a bad method name, but communicates the point
      this.delegate.canPreloadAnchor(element)
    }
  }
}
```

Then, the `Session` instance could build an observer (and manage its `.start()` and `.stop()` calls with [the rest of its observers](https://github.com/hotwired/turbo/blob/c207f5b25758e4a084e8ae42e49712b91cf37114/src/core/session.js#L27-L35):

```js
export class Session {
  // ...
  preloadObserver = new PreloadObserver(this, document.body)

  start() {
    // ...
    this.preloadObserver.start()
    // ...
  }

  stop() {
    // ...
    this.preloadObserver.stop()
    // ...
  }

  // Preload observer delegate

  canPreloadAnchor(anchor) {
    this.preloader.preloadURL(anchor)
  }
}
```

Reply:
I very much like this suggestion, but please forgive me if I don't lift a finger until I hear back from one of the maintainers.